### PR TITLE
Add visual studio code editor

### PR DESCRIPTION
**Pull request by mucaho**
This is the 3rd implementation variant as proposed in #14354. See that issue for details:
- [x] Still need to adapt the configuration file, which enables a vital part of the application - the extension manager. Seems like it can stay FOSS even with the configuration file pointing at official marketplace.
- [x] This PR downgrades current version of `electron` to be compatible with `vscode`. `electron` will have to follow `vscode`s version requirement in future. TODO: make seperate `electron` nix expression, just for `vscode`.
- [x] This PR shifts the injection of shared library lookup paths from the `wrapProgram` utility to `patchELF` itself. The flag `dontPatchELF` is needed, because in the `fixupPhase` the added `$out/bin` rpath is automatically removed. Additionally, both `dontPatchELF` & `dontStrip` need to be set to `false` in order for the binary to run without segfault. TODO: Move `patchELF` to `postFixup`.
###### Things done:
- [x] Tested using buildtime sandboxing (`nix-build . --option build-use-chroot true -A vscode`)
- [x] Tested using runtime sandboxing (`nix-shell -I nixpkgs=/path/to/vscode/branch --pure -p vscode`)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip --against HEAD~2"` from root of nixpkgs fork
  - [x] dependent pkg `electron` compiles & runs
  - [x] dependent pkg `vscode` compiles & runs
  - [x] Tested using `electron-quick-start`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More

Fixes issue #14354 

cc @travisbhartwell

---

### CODE REVIEW

**Comment by vcunat** [comment]
> The flag dontPatchELF is needed, because in the fixupPhase the added $out/bin rpath is automatically removed. Additionally, both dontPatchELF & dontStrip need to be set to false in order for the binary to run without segfault.

It's generally better to run `patchelf` in `postFixup`. That way you shouldn't need to disable the automating stripping and patchelfing. (Stripping a patchelfed file breaks it sometimes.)

**Comment by vcunat** [comment]
And :+1: for using RPATH instead of $LD_LIBRARY_PATH, as the latter e.g. "relinks" any child processes, etc.

File: pkgs/applications/editors/vscode/default.nix [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        cp ./product.json "$out"/product.json
+        cp ./tslint.json "$out"/tslint.json
+        # COPY LEGAL STUFF
+        cp ./LICENSE.txt "$out"/LICENSE.txt
+        cp ./OSSREADME.json "$out"/OSSREADME.json
+        cp ./ThirdPartyNotices.txt "$out"/ThirdPartyNotices.txt
+      '';
+
+      meta = with stdenv.lib; {
+        description = "Code Editing. Redefined.";
```
Comment (jagajaga) [inline pkgs/applications/editors/vscode/default.nix]: Please write a more appreciate description.

File: pkgs/applications/editors/vscode/default.nix [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+          Visual Studio Code is a new type of tool that combines the simplicity of a code editor with what developers need for their core edit-build-debug cycle.
+          Code provides comprehensive editing and debugging support, an extensibility model, and lightweight integration with existing tools.
+        '';
+        homepage = http://code.visualstudio.com/;
+        license = licenses.mit;
+        platforms = [ "x86_64-linux" ];
+      };
+    };
+  in
+    writeScriptBin "vscode" ''
```
Comment (jagajaga) [inline pkgs/applications/editors/vscode/default.nix]: We prefer to use `wrapProgram`.

Comment (mucaho) [inline pkgs/applications/editors/vscode/default.nix]: There is no executable to wrap around, I need to create one from scratch. How can I do that with `wrapProgram`?

Comment (vcunat) [inline pkgs/applications/editors/vscode/default.nix]: That's done by invoking `makeWrapper` directly.

Comment (mucaho) [inline pkgs/applications/editors/vscode/default.nix]: Aha, one can use the lower-level `makeWrapper`, which does not necessarily replace the original executable.

**Comment by mucaho** [comment]
Ok, thank you all for your input, I have applied the suggested changes.

**Comment by ledettwy** [comment]
Why was vscode put into the "unfree" section of nixpkgs?  What are the non-free components in it?

**Comment by gebner** [comment]
@ledettwy The binary distribution of vscode is non-free.  The source distribution as hosted on github is licensed under MIT, but it cannot access the marketplace (i.e., the extensions) out of the box.  It is not clear to me what would happen if we were to patch the open-source version to access the marketplace (you just need to modify a single JSON file).

The original vscode derivation (this PR) was built from source, but it is fundamentally broken since it needs to access the network during the build (#14447).  The current tracking issue for the from-source build is #26698.

**Comment by ledettwy** [comment]
>  It is not clear to me what would happen if we were to patch the open-source version to access the marketplace (you just need to modify a single JSON file).

Not sure what you mean.  Free software that connects to non-free servers doesn't stop becoming free software.  Browsers can have the capability to download extensions from the Internet and the browser is still free.

If you're concerned about the JSON file itself not being free software, doing a clean room implementation of that seems like it would be pretty straightforward.

**Comment by orivej** [comment]
The [license](https://code.visualstudio.com/license) of the binary build that we use says that "You may not … share, publish, or lend the software". This makes it unfree nonredistributable.

**Comment by ledettwy** [comment]
Oh, so it's the binary version that's in unfree nixpkgs, and the source version (when it works) will go into the free nixpkgs?

**Comment by orivej** [comment]
Yes.

**Pull request merged**